src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Vectorized.cs

```diff
@@ -0,0 +1,201 @@
+            while (length >= Vector128<byte>.Count * 4)
+            {
+                x5 = CarrylessMultiplyLower(x1, x0);
+                Vector128<ulong> x6 = CarrylessMultiplyLower(x2, x0);
+                Vector128<ulong> x7 = CarrylessMultiplyLower(x3, x0);
+                Vector128<ulong> x8 = CarrylessMultiplyLower(x4, x0);
+
+                x1 = CarrylessMultiplyUpper(x1, x0);
+                x2 = CarrylessMultiplyUpper(x2, x0);
+                x3 = CarrylessMultiplyUpper(x3, x0);
+                x4 = CarrylessMultiplyUpper(x4, x0);
+
+                Vector128<ulong> y5 = Vector128.LoadUnsafe(ref srcRef).AsUInt64();
+                Vector128<ulong> y6 = Vector128.LoadUnsafe(ref srcRef, 16).AsUInt64();
+                Vector128<ulong> y7 = Vector128.LoadUnsafe(ref srcRef, 32).AsUInt64();
+                Vector128<ulong> y8 = Vector128.LoadUnsafe(ref srcRef, 48).AsUInt64();
+
+                x1 ^= x5;
+                x2 ^= x6;
+                x3 ^= x7;
+                x4 ^= x8;
+
+                x1 ^= y5;
+                x2 ^= y6;
+                x3 ^= y7;
+                x4 ^= y8;
+
+                srcRef = ref Unsafe.Add(ref srcRef, Vector128<byte>.Count * 4);
+                length -= Vector128<byte>.Count * 4;
+            }
+
+            // Fold into 128-bits.
+            x0 = Vector128.Create(0x01751997d0, 0x00ccaa009e).AsUInt64(); // k3, k4
+
+            x5 = CarrylessMultiplyLower(x1, x0);
+            x1 = CarrylessMultiplyUpper(x1, x0);
+            x1 ^= x2;
+            x1 ^= x5;
+
+            x5 = CarrylessMultiplyLower(x1, x0);
+            x1 = CarrylessMultiplyUpper(x1, x0);
+            x1 ^= x3;
+            x1 ^= x5;
+
+            x5 = CarrylessMultiplyLower(x1, x0);
+            x1 = CarrylessMultiplyUpper(x1, x0);
+            x1 ^= x4;
+            x1 ^= x5;
+
+            // Single fold blocks of 16, if any.
+            while (length >= Vector128<byte>.Count)
+            {
+                x2 = Vector128.LoadUnsafe(ref srcRef).AsUInt64();
+
+                x5 = CarrylessMultiplyLower(x1, x0);
+                x1 = CarrylessMultiplyUpper(x1, x0);
+                x1 ^= x2;
+                x1 ^= x5;
+
+                srcRef = ref Unsafe.Add(ref srcRef, Vector128<byte>.Count);
+                length -= Vector128<byte>.Count;
+            }
+
+            // Fold 128 bits to 64 bits.
+            x2 = CarrylessMultiplyLeftLowerRightUpper(x1, x0);
+            x3 = Vector128.Create(~0, 0, ~0, 0).AsUInt64();
+            x1 = Sse2.IsSupported
+                ? Sse2.ShiftRightLogical128BitLane(x1, 8)
+                : AdvSimd.ExtractVector128(x1.AsByte(), Vector128<byte>.Zero, 8).AsUInt64();
+            x1 ^= x2;
+
+            x0 = Vector128.CreateScalar(0x0163cd6124).AsUInt64(); // k5, k0
+
+            x2 = Sse2.IsSupported
+                ? Sse2.ShiftRightLogical128BitLane(x1, 4)
+                : AdvSimd.ExtractVector128(x1.AsByte(), Vector128<byte>.Zero, 4).AsUInt64();
```

Comment:
Should this be extracted to a (local) helper-function?
* with proper funciton name it's self-documenting
* the repeation from L171-L173 can be removed